azure-iot-device/azure/iot/device/provisioning/internal/request_response_provider.py

```diff
@@ -21,74 +21,73 @@ class RequestResponseProvider(object):
 
     def disconnect(self, callback=None):
         if callback is None:
             callback = self._on_connection_state_change
-        self._state_based_provider.disconnect(callback=callback)
+        self._mqtt_transport.disconnect(callback=callback)
 
     def publish(self, topic, request, callback=None):
         if callback is None:
             callback = self._on_publish_completed
-        self._state_based_provider.publish(topic=topic, message=request, callback=callback)
+        self._mqtt_transport.publish(topic=topic, message=request, callback=callback)
 
-    def subscribe(self, topic, callback=None):
+    def enable_responses(self, callback=None):
         if callback is None:
             callback = self._on_subscribe_completed
-        self._state_based_provider.subscribe(topic=topic, callback=callback)
+        self._mqtt_transport.enable_responses(callback=callback)
 
-    def unsubscribe(self, topic, callback=None):
+    def disable_responses(self, callback=None):
         if callback is None:
             callback = self._on_unsubscribe_completed
-        self._state_based_provider.unsubscribe(topic=topic, callback=callback)
+        self._mqtt_transport.disable_responses(callback=callback)
 
-    def _receive_response(self, topic, payload):
+    def _receive_response(self, rid, status_code, key_value_dict, response):
         """
         Handler that processes the response from the service.
-        :param topic: The topic in bytes name on which the message arrived on.
-        :param payload: Payload in bytes of the message received.
+        :param rid: The id of the request which is being responded to.
+        :param status_code: The status code inside the response
+        :param key_value_dict: A dictionary of keys mapped to a list of values extracted from the topic of the response.
+        :param response: String payload of the message received.
+        :return:
         """
         # """ Sample topic and payload
         # $dps/registrations/res/200/?$rid=28c32371-608c-4390-8da7-c712353c1c3b
         # {"operationId":"4.550cb20c3349a409.390d2957-7b58-4701-b4f9-7fe848348f4a","status":"assigning"}
         # """
-        logger.info("Received payload:")
-        logger.info(payload)
-        # topic_str = topic.decode("utf-8")
-        if payload is not None:  # In cases of empty erroneous response from service
-            response = payload.decode("utf-8")
-
-        # There may be no response when status code is >= 300
-        logger.info("Received response:{} on topic:{}".format(response, topic))
-
-        if topic.startswith("$dps/registrations/res/"):
-            topic_parts = topic.split("$")
-            key_value_dict = urllib.parse.parse_qs(topic_parts[POS_QUERY_PARAM_PORTION])
-            rid = key_value_dict["rid"][0]
-        # TODO Not DPS may have other ways to retrieve rid
+        logger.info("Received response {}:".format(response))
 
         if rid in self._pending_requests:
```

Comment:
>if rid in self._pending_requests: [](start = 8, length = 33)

Is the management of pending requests threadsafe?

This design of how pending requests are managed resembles the model used down in the MQTTProvider, and I just had to fix a threading bug there. You may find that something similar will arise here - I'd check out the latest changes to the MQTTProvider in master branch and see if what's going on there applies here.

Reply:
I don't think there is any issue is similar to the one mqtt provider has. The pending requests have the request or does not have it. After seeing Bert's twin PR today I am thinking what to do if the rid is not there. But that is a different discussion point.

---
In reply to: [287547224](https://github.com/Azure/azure-iot-sdk-python-preview/pull/97#discussion_r287547224) [](ancestors = 287547224)